File: backend/app/api/src/main/kotlin/com/wallet/clover/api/controller/LottoSpotController.kt
package com.wallet.clover.api.controller

import com.wallet.clover.api.common.CommonResponse
import com.wallet.clover.api.common.PageResponse
import com.wallet.clover.api.dto.LottoSpot
import com.wallet.clover.api.dto.LottoWinningStore
import com.wallet.clover.api.service.LottoSpotService
import com.wallet.clover.api.service.LottoWinningStoreService
import kotlinx.coroutines.flow.Flow
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/lotto-spots")
class LottoSpotController(
    private val lottoSpotService: LottoSpotService,
    private val lottoWinningStoreService: LottoWinningStoreService
) {

    @GetMapping
    suspend fun getAllLottoSpots(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int
    ): CommonResponse<PageResponse<LottoSpot.Response>> {
        return CommonResponse.success(lottoSpotService.getAllLottoSpots(page, size))
    }

    @GetMapping("/search")
    suspend fun searchByName(@RequestParam name: String): CommonResponse<List<LottoSpot.Response>> {
        return CommonResponse.success(lottoSpotService.searchByName(name))
    }

    @PostMapping("/crawl/{round}")
    suspend fun crawlStores(@PathVariable round: Int): CommonResponse<String> {
        lottoWinningStoreService.crawlWinningStores(round)
        return CommonResponse.success("Crawling started for round ")
    }

    @GetMapping("/winning/{round}")
    suspend fun getWinningStores(@PathVariable round: Int): CommonResponse<List<LottoWinningStore.Response>> {
        val stores = lottoWinningStoreService.getWinningStores(round)
        return CommonResponse.success(stores.map { LottoWinningStore.Response.from(it) })
    }

    @GetMapping("/{spotId}/history")
    suspend fun getSpotWinningHistory(@PathVariable spotId: Long): CommonResponse<List<LottoWinningStore.Response>> {
        val spot = lottoSpotService.getSpotById(spotId)
        val history = lottoWinningStoreService.getWinningHistoryByName(spot.name)
        return CommonResponse.success(history.map { LottoWinningStore.Response.from(it) })
    }
}

File: backend-node/src/lotto-spot/lotto-spot.controller.ts
import { Controller, Get, Post, Param, Query } from '@nestjs/common';
import { LottoSpotService } from './lotto-spot.service';
import { LottoWinningStoreService } from './lotto-winning-store.service';
import { WinningInfoCrawlerService } from '../lotto/winning-info-crawler.service';

@Controller('lotto-spots')
export class LottoSpotController {
    constructor(
        private readonly lottoSpotService: LottoSpotService,
        private readonly lottoWinningStoreService: LottoWinningStoreService,
        private readonly winningInfoCrawlerService: WinningInfoCrawlerService
    ) { }

    @Get()
    async getAllLottoSpots(
        @Query('page') page = 0,
        @Query('size') size = 20
    ) {
        return this.lottoSpotService.getAllLottoSpots(+page, +size);
    }

    @Get('search')
    async searchByName(@Query('name') name: string) {
        return this.lottoSpotService.searchByName(name);
    }

    @Post('crawl/:round')
    async crawlStores(@Param('round') round: string) {
        const result = await this.lottoWinningStoreService.crawlWinningStores(+round);
        return { message: `Crawling started for round ${round}`, result };
    }

    @Post('winning-info/:round')
    async crawlWinningInfo(@Param('round') round: string) {
        await this.winningInfoCrawlerService.crawlWinningInfo(+round);
        return { message: `Winning Info Crawling started for round ${round}` };
    }

    @Get('winning-info/:round')
    async getWinningInfo(@Param('round') round: string) {
        return this.winningInfoCrawlerService.getWinningInfo(+round);
    }

    @Get('winning/:round')
    async getWinningStores(@Param('round') round: string) {
        return this.lottoWinningStoreService.getWinningStores(+round);
    }

    @Get(':spotId/history')
    async getSpotWinningHistory(@Param('spotId') spotId: string) {
        const spot = await this.lottoSpotService.getSpotById(BigInt(spotId));
        if (!spot) {
            throw new Error("Spot not found");
            // In real NestJS, use NotFoundException. Will fix if compilation complains or standard requires it.
        }
        return this.lottoWinningStoreService.getWinningHistoryByName(spot.name);
    }
}
